xmtp/src/storage/encrypted_store/mod.rs

```diff
@@ -360,6 +360,40 @@ impl EncryptedMessageStore {
             .execute(conn)?;
         Ok(())
     }
+
+    pub fn fetch_and_lock_outbound_payloads(
+        &self,
+        payload_state: OutboundPayloadState,
+        lock_duration_ns: i64,
+    ) -> Result<Vec<StoredOutboundPayload>, StorageError> {
+        let conn = &mut self.conn()?;
+        use self::schema::outbound_payloads::dsl::*;
+        let now = now();
+        // Must happen atomically
+        let payloads = diesel::update(outbound_payloads)
+            .filter(outbound_payload_state.eq(payload_state as i32))
+            .filter(locked_until_ns.lt(now))
+            .set(locked_until_ns.eq(now + lock_duration_ns))
```

Comment:
One thing we'll have to be mindful of with this approach is that we need to somehow retrigger this process if rows have been excluded because of the `locked_until_ns`. I wouldn't expect this to be running in a loop or on a cron schedule, but we'll somehow need to make sure this gets re-run in some way other than the user sending another message.